src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs

```diff
@@ -134,5 +127,268 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+                                    conversionDiagnostics.Clear();
+                                    reported = true;
+                                }
+                            }
+
+                            if (!reported)
+                            {
+                                CompoundUseSiteInfo<AssemblySymbol> useSiteInfo = GetNewCompoundUseSiteInfo(conversionDiagnostics);
+                                var conversion = Conversions.ClassifyConversionFromExpression(value, objectType, ref useSiteInfo);
+                                if (!conversion.Exists || !conversion.IsValid || useSiteInfo.HasErrors)
+                                {
+                                    CheckFeatureAvailability(value.Syntax, MessageID.IDS_FeatureImprovedInterpolatedStrings, diagnostics);
+                                }
+                            }
+                        }
+
+                        newPart = insert.Update(newPart, insert.Alignment, insert.Format, newPart.Type);
+                    }
+                    else
+                    {
+                        Debug.Assert(newPart is BoundLiteral { ConstantValue: { IsString: true } });
+                        Debug.Assert(currentPart.ConstantValue is { IsString: true });
+                        baseStringLength += currentPart.ConstantValue.RopeValue!.Length;
+                    }
+
+                    partsBuilder.Add(newPart);
+
+                    ReportDiagnosticsIfObsolete(diagnostics, currentFormatCall.Method, currentPart.Syntax, hasBaseReceiver: false);
+                    ReportDiagnosticsIfUnmanagedCallersOnly(diagnostics, currentFormatCall.Method, currentPart.Syntax.Location, isDelegateConversion: false);
+                }
+
+                conversionDiagnostics?.Free();
+
+                var newParts = partsBuilder.ToImmutableAndFree();
+
+                var createMethod = (MethodSymbol)GetWellKnownTypeMember(
+                    WellKnownMember.System_Runtime_CompilerServices_InterpolatedStringBuilder__CreateInt32Int32,
+                    diagnostics,
+                    syntax: unconvertedInterpolatedString.Syntax);
+
+                MethodArgumentInfo? constructorInfo = null;
+
+                if (createMethod != null)
+                {
+                    var intType = GetSpecialType(SpecialType.System_Int32, diagnostics, unconvertedInterpolatedString.Syntax);
+                    var arguments = ImmutableArray.Create<BoundExpression>(
+                        new BoundLiteral(unconvertedInterpolatedString.Syntax, ConstantValue.Create(baseStringLength), intType) { WasCompilerGenerated = true },
+                        new BoundLiteral(unconvertedInterpolatedString.Syntax, ConstantValue.Create(numFormatHoles), intType) { WasCompilerGenerated = true });
+
+                    constructorInfo = new MethodArgumentInfo(createMethod, arguments, ArgsToParamsOpt: default, DefaultArguments: BitVector.Empty, Expanded: false);
```

Comment:
>constructorInfo = new MethodArgumentInfo(createMethod, arguments, ArgsToParamsOpt: default, DefaultArguments: BitVector.Empty, Expanded: false); [](start = 20, length = 144)

I am not sure why is it necessary to do this work in binding, we could easily synthesize the call in lowering.

Reply:
This will need to get more generic to support non-well-known create methods for other builder types. I _could_ special case this instance to not do the work now and do it in lowering instead, but that will just result in more code dealing with the special case down the road and I'd rather avoid it.

---
In reply to: [610915127](https://github.com/dotnet/roslyn/pull/52422#discussion_r610915127) [](ancestors = 610915127)

Reply:
> This will need to get more generic to support non-well-known create methods for other builder types. I could special case this instance to not do the work now and do it in lowering instead, but that will just result in more code dealing with the special case down the road and I'd rather avoid it.

Then I would consider not using WellKnownMembers at all, but rather do the normal binding for a call.

---
In reply to: [612042647](https://github.com/dotnet/roslyn/pull/52422#discussion_r612042647) [](ancestors = 612042647,610915127)

Reply:
I don't think this is necessary. This codepath is only ever dealing with one type, and that type needs to have the things we expect or it's malformed. Normal binding does a lot of work we don't care about here.

---
In reply to: [612777789](https://github.com/dotnet/roslyn/pull/52422#discussion_r612777789) [](ancestors = 612777789,612042647,610915127)